scene/2d/skeleton_2d.cpp

```diff
@@ -77,8 +170,198 @@ void Bone2D::_notification(int p_what) {
 			skeleton = nullptr;
 		}
 		parent_bone = nullptr;
+		set_transform(cache_transform);
+	}
+
+	if (p_what == NOTIFICATION_READY) {
+		if (autocalculate_length_and_angle) {
+			calculate_length_and_rotation();
+		}
+	}
+
+#ifdef TOOLS_ENABLED
+	if (p_what == NOTIFICATION_EDITOR_PRE_SAVE || p_what == NOTIFICATION_EDITOR_POST_SAVE) {
+		Transform2D tmp_trans = get_transform();
+		set_transform(cache_transform);
+		cache_transform = tmp_trans;
+	}
+
+	// Bone2D Editor gizmo drawing:
+	if (p_what == NOTIFICATION_DRAW) {
+		// Only draw the gizmo in the editor!
+		if (Engine::get_singleton()->is_editor_hint() == false) {
```

Comment:
Usually, the gizmo code goes inside some editor plug-in; while, inside the nodes we have the gizmos that we want to see at runtime.
For example the physical shape, all have the resizeable gizmo handled by the [CollisionShape3DGizmoPlugin](https://github.com/godotengine/godot/blob/master/editor/node_3d_editor_gizmos.cpp#L3510), but the gizmo used to show the shapes at runtime is handled by the [shape resource itself](https://github.com/godotengine/godot/blob/master/scene/resources/box_shape_3d.cpp#L34).

Reply:
Yeah, I asked Joan about this at the time and I don't remember if I was told one way or another, but I know there was not objection. The old gizmo system for 2D bones was overly complicated and confusing because of the 'fake' CanvasItem bones, so I removed the 2D bone drawing gizmo and placed it in the Bone2D. It was also in a huge file that handled a ton of 2D gizmos, so moving it out and into the Skeleton2D node seemed cleaner and easier to maintain.

Personally, I think this makes more sense for it to be in the Bone2D. Writing a new gizmo plugin was outside of the scope and time constraints of the GSoC program, but there was bugs with it showing the IK results correctly, so I opted for adding it to the Bone2D directly to fix these issues. A bonus of this is that it fixed the issue with the Bone2D gizmos overlapping on top of sprites, making it hard (if not impossible) to see the sprite underneath. It also allowed for adding the ability to selectively hide the gizmo on bones.

I'm not sure there is really a need to place the gizmo in an editor plugin, especially since the Bone2D should be the only node that needs to have a 2D bone gizmo. I think the benefits of having it in the Bone2D code outweigh the cons.

Reply:
I don't want to delay the merge of this PR, though, this would be an exception to the Godot engine design. When you release the game, you don't need to include the editor gizmos. The gizmo, are integrated via plugin to correctly separate the two functionalities.
But, also, are you sure that this code already compiles without the editor in release mode?

Reply:
I think there are two options, at least from my perspective:

The first option: Keep the gizmo code as it is currently and (potentially) change it at a later date.
* The code only runs in the editor and will not affect exported games (that's what checking for the editor hint is for) so there is little for end users.
* I would argue that its more readable than the previous solution, where the code for Bone2D gizmos were in a huge file. It is not necessarily the ideal solution, but it is at least more readable.
* It is, I guess, harder to follow, but at least all the code for Bone2D is in one place, rather than spread across multiple files (without documentation on where to look BTW)
* This would also have all the functional benefits: per-Bone2D gizmo visibility, the gizmo facing in the correct bone angle (especially on tip bones), the color changing when IK is executing, and the bones gizmo scale not being constant regardless of editor zoom.

The second option: Put the old gizmo code back.
* This would comply with the normal Godot engine design, as it is what was being used before.
* Its not quite as readable, but it is the old and familiar code
* The difference between editor-only code and scene code is more precisely defined due to them being in separate files.
* Would miss out on the benefits mentioned above
* May not work with the Bone2D node adjustments made in the rest of the PR

It's up to you guys, I can do it either way.

Reply:
But why you can't create a new plugin and move the code in there? It seems doable and even easier.

Reply:
I don’t know where to write the plugin, which class to use, etc. If there is an example or something I can refer to, I do not mind taking a look and seeing about possibly using that instead, especially if it’s easier.

Reply:
Here you can find some examples:
- `CollisionShape` gizmo: https://github.com/godotengine/godot/blob/master/editor/node_3d_editor_gizmos.cpp#L3510
- `Skeleton` gizmo: https://github.com/godotengine/godot/blob/master/editor/node_3d_editor_gizmos.cpp#L1550
You can find many other in there.